src/PostgREST/App.hs

```diff
@@ -93,22 +101,26 @@ app dbStructure conf apiRequest =
         (ActionRead, TargetIdent qi, Nothing) ->
           case readSqlParts of
             Left errorResponse -> return errorResponse
-            Right (q, cq) -> do
-              let stm = createReadStatement q cq (contentType == CTSingularJSON) shouldCount (contentType == CTTextCSV)
-              row <- H.query () stm
-              let (tableTotal, queryTotal, _ , body) = row
-                  (status, contentRange) = rangeHeader queryTotal tableTotal
-                  canonical = iCanonicalQS apiRequest
-              return $
-                if contentType == CTSingularJSON && queryTotal /= 1
-                  then singularityError (toInteger queryTotal)
-                  else responseLBS status
-                    [toHeader contentType, contentRange,
-                      ("Content-Location",
-                        "/" <> toS (qiName qi) <>
-                          if BS.null canonical then "" else "?" <> toS canonical
-                      )
-                    ] (toS body)
+            Right (q, cq) -> 
```

Comment:
One idea to reduce the nesting of these cases:

```haskell
(ActionRead, TargetIdent qi, Nothing) ->
  let partsFields = (,) <$> readSqlParts
        <*> (binaryField contentType =<< fldNames) in
  case partsFields of
    Left errorResponse -> return errorResponse
    Right ((q, cq), bField) -> do
```

Sadly it is more dense and symbolic and introduces another variable name but it does prevent nested case statements dealing with Either.